### PR TITLE
Tile variants content

### PR DESCRIPTION
**Pull request by moonheart08**
## About the PR
Requires https://github.com/space-wizards/RobustToolbox/pull/2577
**Screenshots**
image

**Changelog**
:cl: moony
- add: Some tiles now have additional visual variants. These will start showing up when maps are updated.

### CODE REVIEW

**Comment by moonheart08** [comment]
image

**Comment by Peptide90** [comment]
You had mentioned not using other tile variants (I've used some of the old asteroid ones).

Will this change make the map not load or can it internet the tiles still as variants?

**Comment by moonheart08** [comment]
> You had mentioned not using other tile variants (I've used some of the old asteroid ones).
> 
> Will this change make the map not load or can it interpret the tiles still as variants?

It'll keep working, I'll write a migration tool before I consider making maps no longer load.

File: Content.Server/Administration/Commands/VariantizeCommand.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+[AdminCommand(AdminFlags.Mapping)]
+public sealed class VariantizeCommand : IConsoleCommand
+{
+
+    public string Command => "variantize";
+
+    public string Description => "Automatic variant randomization.";
+
+    public string Help => "variantize <grid id>";
```
Comment (mirrorcult) [inline Content.Server/Administration/Commands/VariantizeCommand.cs:17]: we're still localizing these i think

File: Resources/Prototypes/Tiles/floors.yml [hunk 2]
```diff
@@ -32,6 +36,8 @@
   id: floor_white
   name: white floor
   texture: "white"
+  variants: 4
+  placementVariants: [0, 1, 2, 3]
```
Comment (mirrorcult) [inline Resources/Prototypes/Tiles/floors.yml:40]: whats the point of `placementVariants` exactly, why cant it just be generated from the # of variants

**Review by mirrorcult — CHANGES_REQUESTED** [review]

File: Resources/Prototypes/Tiles/floors.yml (hunk 2 above)
Comment (moonheart08) [reply Resources/Prototypes/Tiles/floors.yml:40]: > whats the point of `placementVariants` exactly, why cant it just be generated from the # of variants

tl;dr variants may not always be for just simple mapping visuals so I thought i'd plan ahead.

Comment (mirrorcult) [reply Resources/Prototypes/Tiles/floors.yml:40]: make sense

File: Content.Server/Administration/Commands/VariantizeCommand.cs [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+    public string Description => Loc.GetString("variantize-command-description");
+
+    public string Help => Loc.GetString("variantize-command-help-text");
+
+    public void Execute(IConsoleShell shell, string argStr, string[] args)
+    {
+        if (args.Length != 1)
+        {
+            shell.WriteLine(Loc.GetString("shell-wrong-arguments-number"));
```
Comment (metalgearsloth) [inline Content.Server/Administration/Commands/VariantizeCommand.cs:23]: writeerror

File: Content.Server/Administration/Commands/VariantizeCommand.cs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+            return;
+        }
+
+        var mapManager = IoCManager.Resolve<IMapManager>();
+        var random = IoCManager.Resolve<IRobustRandom>();
+
+        if (!int.TryParse(args[0], out var targetId))
+        {
+            shell.WriteLine(Loc.GetString("shell-argument-must-be-number"));
```
Comment (metalgearsloth) [inline Content.Server/Administration/Commands/VariantizeCommand.cs:32]: writeerror

File: Content.Server/Administration/Commands/VariantizeCommand.cs [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+            return;
+        }
+
+        var gridId = new GridId(targetId);
+        var grid = mapManager.GetGrid(gridId);
+        foreach (var tile in grid.GetAllTiles())
```
Comment (metalgearsloth) [inline Content.Server/Administration/Commands/VariantizeCommand.cs:38]: Should also use TryGetGrid as well.

File: Content.Shared/Maps/ContentTileDefinition.cs [hunk 6]
```diff
@@ -40,6 +40,10 @@ public sealed class ContentTileDefinition : IPrototype, IInheritingPrototype, IT
 
         [DataField("friction")] public float Friction { get; set; }
 
+        [DataField("variants")] public byte Variants { get; set; } = 1;
+
+        [DataField("placementVariants")] public List<byte> PlacementVariants { get; set; } = new() { 0 };
```
Comment (metalgearsloth) [inline Content.Shared/Maps/ContentTileDefinition.cs:45]: Shouldn't variants just use the length of placementvariants?

Also should placementvariants just be an array?

Comment (moonheart08) [reply Content.Shared/Maps/ContentTileDefinition.cs:45]: No and yes. Some tiles may have no random placement variants but multiple variant sprites. PlacementVariants is specifically "what variants can be used by the `variantize` command."

Comment (metalgearsloth) [reply Content.Shared/Maps/ContentTileDefinition.cs]: Should have a doc comment explaining it then.

**Review by PaulRitter — APPROVED** [review]
